### ISSUE TITLE
Use FFmpeg for video capture

### ISSUE THREAD

**Issue by weirddan455** [dosbox-staging/dosbox-staging#2764]
### Are you using the latest Dosbox-Staging Version?

- [X] I have checked releases and am using the latest release.

### What Operating System are you using?

Linux x86_64

### If Other OS, please describe

Arch Linux

### Is your feature request related to a problem? Please describe.

Not strictly related but I saw this open ticket for video capture #2720

### Describe the solution you'd like

Currently, video capture is using ZMBV codec which I had never heard of before.  VLC (and some Google searches) actually calls it "DosBox Capture Codec" which makes me think we're one of, if not the only one using this thing.

I'm considering writing support for FFmpeg as another encoding solution.  Maybe a `VideoEncoder` interface class would be good so we can choose between encoders at runtime.  Looking at `src/capture/capture_video.h`, it might just be 3 virtual functions needed.  Then wrap current code into a ZMBV class and write a new one for FFmpeg.  I think it wouldn't be too hard (famous last words...) and I have written a small application using FFmepg in the past.

The main advantages to this would be:

1. Support for standard codecs with better compression.  FFMpeg supports a ton of them and it could be user configurable.  Something like H.264 would make a good default I think.

2. Support for GPU accelerated real-time encoding.  This does depend on your GPU hardware supporting the specific format you're trying to output but it can always fall back to CPU encoding.  Support for this is baked into the library.

### Describe alternatives you've considered

I don't really know of a better video encoding library.  FFmpeg is very widely used and cross platform.  But if anyone knows of any alternatives, let me know.

### Code of Conduct & Contributing Guidelines

- [X] Yes, I agree.

**Comment by johnnovak** [comment]
It think this would be a welcome addition @weirddan455. Many times I think twice before bothering with a video capture if I need to run the resulting ZMBV capture through `ffmpeg` manually. Too fiddly, and I'm old 😅 

Non-ancient computers should be able to handle GPU-accelerated MPEG encoding in realtime just fine. Just found this one, this guy did something similar with NVENC/H.264/HEVC:
https://gist.github.com/Brainiarc7/4636a162ef7dc2e8c9c4c1d4ae887c0e

ZMBV should continue to be an option, because why not, and on low-powered devices that's the only option to capture the raw output with minimal zlib/delta compression. Probably you've found this already, but here we go:
https://wiki.multimedia.cx/index.php/DosBox_Capture_Codec

I know some people use ZMBV as a plugin for WinUAE to capture raw, lossless footage of their emualted Amiga. Most other codecs introduce stuttering/lags into WinUAE (probably the capture is not threaded).

More generally, the ability to capture video in lossless manner should continue to be an option. H.264 can do lossless AFAIK. Just did some random searching on the topic, and found out that with NVENC/x264 lossy compression that's actually "near indistinguishable" from lossless it's also a good option to hit that performance/quality/filesize sweet spot:

> To get NEARLY lossless footage with x264 would be to use the Indistinguishable Quality, Large File Size quality setting and the Software (x264 low CPU usage preset, increases file size) encoder option. 

Anyway, I'll leave the design to you; all I'm saying we should have all these options, and it should be threaded.

Btw, the second issue you linked (https://github.com/dosbox-staging/dosbox-staging/issues/2734) is unrelated to video capturing so you might wanna remove that. It's just about passing the image with the correct aspect ratio (correct dimensions) to the video encoder, so nothing to do with the proposed change (and I'll fix it shortly).

**Comment by weirddan455** [comment]
>  ZMBV should continue to be an option, because why not, and on low-powered devices that's the only option to capture the raw output with minimal zlib/delta compression.

Not too sure about the performance of this codec either.  There was a lot of stuttering in-game when I was testing just now (especially in the audio), though the audio in the capture output was fine.  Maybe running it in a separate thread will fix that issue.  Either way, I wasn't planning on removing it.

Right now I'm just trying to fix ffmpeg link errors...

**Comment by johnnovak** [comment]
> Maybe running it in a separate thread will fix that issue.

Threading is a must, IMO, yeah.

**Comment by weirddan455** [comment]
Got it.  Finally figured out that link error.  After like an hour of messing around with meson, it turns out I just needed to wrap the header files in `extern "C" {}`.  This is the one library that doesn't do that in its own header files...

**Comment by kcgen** [comment]
> Not too sure about the performance of this codec either. There was a lot of stuttering in-game when I was testing just now even in a release build (especially in the audio)

Yup, during capture, the zmbv codec becomes an active module within the 1ms tick loop (including the PCM audio capture), instead of the frames and audio being shunted out to growing queues digested by a threaded codec.

With it in-band, zmbv eats a significant chunk of that 1ms pie, so peak cycles available will go down a lot (when running with auto or max cycles). 

The trick to get good recording with it is to always use fixed cycles, and enough cycles as you would normally use for smooth gameplay. 

The "cost" of this approach is that you need to suffer through very stuttery and slow-mo experience while recording, but inside the DOS realm, the game will get its desired cycles every "emulated millisecond tick" (despite each emulated 1ms of time taking a lot more actual time). But the benefit is that the recorded frames and audio will be "perfect" (you just need to suffer through it as the player 😅). 

It's a rough experience, so any threading and hardware acceleration would be a massive win for users! I could see this being super helpful for users on the pi or orange pi (which itself can record 8k h265 at 60fps, and there are already ffmpeg patches flowing in).

**Comment by weirddan455** [comment]
It's coming along.  I got ffmpeg to produce correct video output with H.264 and no stuttering despite it running single threaded in software on the CPU so ffmpeg performance may end up being better than the current encoder.  It even has support for ZMBV codec though I haven't tried it yet.  I need to get audio working and file headers written out.  Right now I can only get it to play back with the `ffplay` command.  VLC doesn't know what to do with it.

As far as threading goes (once I get to that), I'm pretty sure it still needs synchronization somewhere.  I'm thinking a `std::queue` or something guarded with a mutex.  Main thread takes a mutex lock and copies the rendered frame into the queue.  Encoder thread takes a lock and copies it to a local buffer (which it needs to do anyway) then releases the lock and does its encoding work.

Basically make a thread safe work queue for the encoder.  That's kind of the way I've done threading in the past though I don't have a ton of experience with it admittedly.

**Comment by kcgen** [comment]
The rwqueue in the source tree might be useful. Its a bounded queue that will block the producer when full (unblocking when room is available) and will block the consumer when it runs empty (unblocking when more data is available). 
The blocking lets you avoid mutexes and locks, and prevents the threads on either side from spinning. 
We use it to run fluidsynth and mt32emu in threads; the sysex messages are rwqeue'd, and likewise the audio frames out are also rwqueue'd. @johnnovak's recent threaded screenshots also use it to thread capturing.

**Comment by weirddan455** [comment]
> The rwqueue in the source tree might be useful. Its a bounded queue that will block the producer when full (unblocking when room is available) and will block the consumer when it runs empty (unblocking when more data is available). The blocking lets you avoid mutexes and locks, and prevents the threads on either side from spinning. We use it to run fluidsynth and mt32emu in threads; the sysex messages are rwqeue'd, and likewise the audio frames out are also rwqueue'd. @johnnovak's recent threaded screenshots also use it to thread capturing.

Yeah, that looks perfect.  That was kind of what I was envisioning.  It's just already done and I didn't know it.

**Comment by johnnovak** [comment]
Oh yeah, by all means use rwqueue @weirddan455.

**Comment by weirddan455** [comment]
Still some more work to be done before I put in a PR but I've got the basics down.  Audio is working and it's getting properly muxed into a .mp4 container with H.264 video and AAC audio.  Zero stutter in-game even single threaded on my 10+ year old CPU on a debug build which is more than I can say about the current encoder.  GDB shows that FFmpeg is spinning up its own internal threads at least for the lib x264 video encoder itself so I'm sure that helps.  Also, 320X200 for a typical DOS game is not exactly too many pixels to encode in the first place.

Probably still a good idea to do threading on our end since I know at least re-sampling and re-scaling are jobs that its doing on the main thread regardless of FFmpeg internal threading.  So next up is handling that + adding in some user config options.

Here's Monkey Island 2 intro I just recorded.

https://github.com/dosbox-staging/dosbox-staging/assets/8128047/b9a7a9a3-32f1-40cc-9fbd-f2724a4f7cdf

**Comment by KyodaiKen** [comment]
Is there any update on this topic? I would like to use FLAC to encode the audio and keep ZMBV but put it into matroska to have a fully lossless compressed file output. Would also be nice if matroska gets told the correct aspect ratio in the container.

**Comment by johnnovak** [comment]
> Is there any update on this topic? I would like to use FLAC to encode the audio and keep ZMBV but put it into matroska to have a fully lossless compressed file output. Would also be nice if matroska gets told the correct aspect ratio in the container.

Not yet. Maybe in the future. But very custom use cases such as what you described won't be supported anyway. Power users with niche uses cases would need to render into ZMBV then do the conversion manually.

In practice, 192kbps MP3 is good enough, and maybe 0.001% of the population hears the difference between FLAC and 320kbps MP3.

**Comment by weirddan455** [comment]
Audio output is already lossless. It's uncompressed PCM.

My main motivation for FFMpeg integration was performance issues with ZMBV but that's been resolved by the move from zlib to zlib-ng so I've somewhat lost interest in this.

> Would also be nice if matroska gets told the correct aspect ratio in the container.

Yeah, maybe we should just switch containers for ZMBV from AVI to matroska so we can get proper aspect ratio.
